flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointFailureManager.java

```diff
@@ -0,0 +1,131 @@
+ */
+public class CheckpointFailureManager {
+
+	private final static int UNLIMITED_TOLERABLE_FAILURE_NUMBER = Integer.MAX_VALUE;
+
+	private final int tolerableCpFailureNumber;
+	private final FailJobCallback failureCallback;
+	private final AtomicInteger continuousFailureCounter;
+	private final Set<Long> countedCheckpointIds;
+
+	public CheckpointFailureManager(int tolerableCpFailureNumber, FailJobCallback failureCallback) {
+		checkArgument(tolerableCpFailureNumber >= 0,
+			"The tolerable checkpoint failure number is illegal, " +
+				"it must be greater than or equal to 0 .");
+		this.tolerableCpFailureNumber = tolerableCpFailureNumber;
+		this.continuousFailureCounter = new AtomicInteger(0);
+		this.failureCallback = checkNotNull(failureCallback);
+		this.countedCheckpointIds = ConcurrentHashMap.newKeySet();
+	}
+
+	/**
+	 * Handle checkpoint exception with a handler callback.
+	 *
+	 * @param exception the checkpoint exception.
+	 * @param checkpointId the failed checkpoint id used to count the continuous failure number based on
+	 *                     checkpoint id sequence. In trigger phase, we may not get the checkpoint id when the failure
+	 *                     happens before the checkpoint id generation. In this case, it will be specified a negative
+	 *                      latest generated checkpoint id as a special flag.
+	 */
+	public void handleCheckpointException(CheckpointException exception, long checkpointId) {
+		CheckpointFailureReason reason = exception.getCheckpointFailureReason();
+		switch (reason) {
+			case PERIODIC_SCHEDULER_SHUTDOWN:
+			case ALREADY_QUEUED:
+			case TOO_MANY_CONCURRENT_CHECKPOINTS:
+			case MINIMUM_TIME_BETWEEN_CHECKPOINTS:
+			case NOT_ALL_REQUIRED_TASKS_RUNNING:
+			case CHECKPOINT_SUBSUMED:
+			case CHECKPOINT_COORDINATOR_SUSPEND:
+			case CHECKPOINT_COORDINATOR_SHUTDOWN:
+			case JOB_FAILURE:
+			case JOB_FAILOVER_REGION:
+			//for compatibility purposes with user job behavior
+			case CHECKPOINT_DECLINED_TASK_NOT_READY:
+			case CHECKPOINT_DECLINED_TASK_NOT_CHECKPOINTING:
+			case CHECKPOINT_DECLINED_ALIGNMENT_LIMIT_EXCEEDED:
+			case CHECKPOINT_DECLINED_ON_CANCELLATION_BARRIER:
+			case CHECKPOINT_DECLINED_SUBSUMED:
+			case CHECKPOINT_DECLINED_INPUT_END_OF_STREAM:
+
+			case EXCEPTION:
+			case CHECKPOINT_EXPIRED:
+			case TASK_CHECKPOINT_FAILURE:
+			case TRIGGER_CHECKPOINT_FAILURE:
+			case FINALIZE_CHECKPOINT_FAILURE:
+				//ignore
+				break;
+
+			case CHECKPOINT_DECLINED:
+				//we should make sure one checkpoint only be counted once
+				if (tolerableCpFailureNumber != UNLIMITED_TOLERABLE_FAILURE_NUMBER
+					&& countedCheckpointIds.add(checkpointId)) {
+					continuousFailureCounter.incrementAndGet();
+				}
+
+				break;
+
+			default:
+				throw new FlinkRuntimeException("Unknown checkpoint failure reason : " + reason.name());
+		}
+
+		if (tolerableCpFailureNumber != UNLIMITED_TOLERABLE_FAILURE_NUMBER
```

Comment:
Nit: Now that I see my suggestion in the context, instead of the repeated check on `tolerableCpFailureNumber != UNLIMITED_TOLERABLE_FAILURE_NUMBER`, at least in the current implementation, you could just put this at the very beginning of the method instead:
`if (tolerableCpFailureNumber != UNLIMITED_TOLERABLE_FAILURE_NUMBER) {return;}` because for the case the method never does anything right now.

Reply:
You are right, but I think you mean:

```
//it's equal not non-equal
if (tolerableCpFailureNumber == UNLIMITED_TOLERABLE_FAILURE_NUMBER) {
	return;
}
```